src/analysis/processing/qgsalgorithmmergevector.cpp

```diff
@@ -259,15 +269,19 @@ QVariantMap QgsMergeVectorAlgorithm::processAlgorithm( const QVariantMap &parame
       QgsAttributes destAttributes;
       for ( const QgsField &destField : outputFields )
       {
-        if ( addLayerField && destField.name() == QLatin1String( "layer" ) )
+        // add source layer information
+        if ( addSourceFields )  
```

Comment:
This isn't needed -- addLayerField/addPathField will always be false if addSourceFields is false
```suggestion
```